module_utils/network/junos/config/l3_interfaces/l3_interfaces.py

```diff
@@ -0,0 +1,244 @@
+        return root_node, intf_unit_node
+
+    def _state_replaced(self, want, have):
+        """ The xml generator when state is replaced
+
+        :rtype: A list
+        :returns: the xml necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        intf_xml = []
+        intf_xml.extend(self._state_deleted(want, have))
+        intf_xml.extend(self._state_merged(want, have))
+        return intf_xml
+
+    def _state_overridden(self, want, have):
+        """ The xml generator when state is overridden
+
+        :rtype: A list
+        :returns: the xml necessary to migrate the current configuration
+                  to the desired configuration
+        """
+        intf_xml = []
+        intf_xml.extend(self._state_deleted(have, have))
+        intf_xml.extend(self._state_merged(want, have))
+        return intf_xml
+
+    def _state_merged(self, want, have):
+        """ The xml generator when state is merged
+
+        :rtype: A list
+        :returns: the xml necessary to merge the provided into
+                  the current configuration
+        """
+        intf_xml = []
+        for config in want:
+            root_node, unit_node = self._get_common_xml_node(config['name'])
+            junosutils.build_child_xml_node(unit_node, 'name',
+                                            str(config['unit']))
+            if config.get('ipv4'):
+                self.build_ipaddr_et(config, unit_node)
+            if config.get('ipv6'):
+                self.build_ipaddr_et(config, unit_node, protocol='ipv6')
+            intf_xml.append(root_node)
+        return intf_xml
+
+    def build_ipaddr_et(self, config, unit_node, protocol='ipv4',
+                        delete=False):
+        family = junosutils.build_child_xml_node(unit_node, 'family')
+        inet = 'inet'
+        if protocol == 'ipv6':
+            inet = 'inet6'
+        ip_protocol = junosutils.build_child_xml_node(family, inet)
+        for ip_addr in config[protocol]:
+            if ip_addr['address'] == 'dhcp' and protocol == 'ipv4':
+                junosutils.build_child_xml_node(ip_protocol, 'dhcp')
+            else:
+                ip_addresses = junosutils.build_child_xml_node(
+                    ip_protocol, 'address')
+                junosutils.build_child_xml_node(
+                    ip_addresses, 'name', ip_addr['address'])
+
+    def _state_deleted(self, want, have):
+        """ The xml configuration generator when state is deleted
+
+        :rtype: A list
+        :returns: the xml configuration necessary to remove the current
+                  configuration of the provided objects
+        """
+        intf_xml = []
+        existing_l3_intfs = [l3_intf['name'] for l3_intf in have]
+
+        for config in want:
+            if config['name'] not in existing_l3_intfs:
+                raise Exception('interface is not a valid l3 interface')
```

Comment:
Why is the exception raised here? If the interface does not have any L3 configuration can't it just be ignored. 
eg:
```
- name: Delete JUNOS L3 logical interface
  junos_l3_interfaces:
    config:
      - name: ge-0/0/1
      - name: ge-0/0/2
  state: deleted
```
In the above case the L3 xml with `'delete'='delete'` attribute added to tags that repesent the module argument can be pushed to device. It the L3 config is present it will delete it else device will ignore the operation.

For 
```
- name: Delete JUNOS L3 logical interface
  junos_l3_interfaces:
  state: deleted
```
For all the interfaces issues a delete xml for L3 attributes. If L3 config is present in running datastore it gets deleted or else ignore.